src/dumbo/src/ns.rs

```diff
@@ -229,49 +248,59 @@ impl MmdsNetworkStack {
             .prepare_eth_unsized(buf, ETHERTYPE_ARP)
             .map_err(WriteArpFrameError::Ethernet)?;
 
-        let arp_len = EthIPv4ArpFrame::write_reply(
-            eth_unsized
-                .inner_mut()
-                .payload_mut()
-                .split_at_mut(ETH_IPV4_FRAME_LEN)
-                .0,
-            self.mac_addr,
-            self.ipv4_addr,
-            self.remote_mac_addr,
-            dst_ipv4,
-        )
-        .map_err(WriteArpFrameError::Arp)?
-        .len();
+        if let Some(ipv4_addr) = self.used_ipv4_addr {
+            let arp_len = EthIPv4ArpFrame::write_reply(
+                eth_unsized
+                    .inner_mut()
+                    .payload_mut()
+                    .split_at_mut(ETH_IPV4_FRAME_LEN)
+                    .0,
+                self.mac_addr,
+                ipv4_addr,
+                self.remote_mac_addr,
+                dst_ipv4,
+            ).map_err(WriteArpFrameError::Arp)?.len();
+
+            return Ok(Some(
+                // The unwrap() is safe because arp_len > 0.
+                NonZeroUsize::new(eth_unsized.with_payload_len_unchecked(arp_len).len()).unwrap(),
+            ));
+        }
+
+        // Error since try to write an ARP reply when there is no pending one.
+        Err(WriteArpFrameError::Arp(ArpFrameError::Operation))
 
-        Ok(Some(
-            // The unwrap() is safe because arp_len > 0.
-            NonZeroUsize::new(eth_unsized.with_payload_len_unchecked(arp_len).len()).unwrap(),
-        ))
     }
 
     fn write_packet(&mut self, buf: &mut [u8]) -> Result<Option<NonZeroUsize>, WritePacketError> {
         let mut eth_unsized = self
             .prepare_eth_unsized(buf, ETHERTYPE_IPV4)
             .map_err(WritePacketError::Ethernet)?;
```

Comment:
This can also go inside the `if` body below, no point in preparing the buffer if it's not going to be used.